pkg/controller/controller_test.go

```diff
@@ -3156,3 +3129,257 @@ func TestProvisionFromPVC(t *testing.T) {
 		}
 	}
 }
+
+func TestProvisionWithMigration(t *testing.T) {
+	var requestBytes int64 = 100000
+	var inTreePluginName = "in-tree-plugin"
+
+	deletePolicy := v1.PersistentVolumeReclaimDelete
+	testcases := []struct {
+		name              string
+		provisioner       string
+		annotation        map[string]string
+		expectTranslation bool
+		expectCreate      bool
+		expectErr         bool
+	}{
+		{
+			name:              "provision with migration on",
+			provisioner:       inTreePluginName,
+			annotation:        map[string]string{annStorageProvisioner: driverName},
+			expectCreate:      true,
+			expectTranslation: true,
+		},
+		{
+			name:              "provision without migration for native CSI",
+			provisioner:       driverName,
+			annotation:        map[string]string{annStorageProvisioner: driverName},
+			expectCreate:      true,
+			expectTranslation: false,
+		},
+		{
+			name:        "ignore in-tree PVC when provisioned by in-tree",
+			provisioner: inTreePluginName,
+			annotation:  map[string]string{annStorageProvisioner: inTreePluginName},
+			expectErr:   true,
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Set up test
+			tmpdir := tempDir(t)
+			defer os.RemoveAll(tmpdir)
+			mockController, driver, _, controllerServer, csiConn, err := createMockServer(t, tmpdir)
+			if err != nil {
+				t.Fatal(err)
+			}
+			mockTranslator := NewMockProvisionerCSITranslator(mockController)
+			defer mockController.Finish()
+			defer driver.Stop()
+			clientSet := fakeclientset.NewSimpleClientset()
+			pluginCaps, controllerCaps := provisionCapabilities()
+			csiProvisioner := NewCSIProvisioner(clientSet, 5*time.Second, "test-provisioner",
+				"test", 5, csiConn.conn, nil, driverName, pluginCaps, controllerCaps,
+				inTreePluginName, false, mockTranslator)
+
+			// Set up return values (AnyTimes to avoid overfitting on implementation)
+			if tc.expectTranslation {
```

Comment:
If we don't expect translation but it accidentally gets called, we still want "foo" set on the objects so that we can assert whether it happened.